storagenode/pieces/store.go

```diff
@@ -118,16 +169,171 @@ func (store *Store) Reader(ctx context.Context, satellite storj.NodeID, pieceID
+	defer mon.Task()(&ctx)(&err)
+
+	v1expired, err := store.expirationInfo.GetExpired(ctx, expiredAt, limit)
+	if err != nil {
+		return nil, err
+	}
+	v0expired, err := store.v0PieceInfo.GetExpired(ctx, expiredAt, limit)
+	if err != nil {
+		return nil, err
+	}
+	return append(v1expired, v0expired...), nil
+}
+
+// SetExpiration records an expiration time for the specified piece ID owned by the specified satellite
+func (store *Store) SetExpiration(ctx context.Context, satellite storj.NodeID, pieceID storj.PieceID, expiresAt time.Time) (err error) {
+	return store.expirationInfo.SetExpiration(ctx, satellite, pieceID, expiresAt)
+}
+
+// DeleteFailed marks piece as a failed deletion.
+func (store *Store) DeleteFailed(ctx context.Context, expired ExpiredInfo, when time.Time) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	if expired.InPieceInfo {
+		return store.v0PieceInfo.DeleteFailed(ctx, expired.SatelliteID, expired.PieceID, when)
+	}
+	return store.expirationInfo.DeleteFailed(ctx, expired.SatelliteID, expired.PieceID, when)
+}
+
+// SpaceUsedForPieces returns the disk space used by all local pieces (both V0 and later).
+// Important note: this metric does not include space used by piece headers, whereas
+// storj/filestore/store.(*Store).SpaceUsed() includes all space used by the blobs.
+func (store *Store) SpaceUsedForPieces(ctx context.Context) (int64, error) {
+	satellites, err := store.getAllStoringSatellites(ctx)
+	if err != nil {
+		return 0, err
+	}
+	var total int64
+	for _, satellite := range satellites {
+		spaceUsed, err := store.SpaceUsedBySatellite(ctx, satellite)
+		if err != nil {
+			return 0, err
+		}
+		total += spaceUsed
+	}
+	return total + store.reservedSpace, nil
+}
+
+func (store *Store) getAllStoringSatellites(ctx context.Context) ([]storj.NodeID, error) {
+	namespaces, err := store.blobs.GetAllNamespaces(ctx)
+	if err != nil {
+		return nil, err
+	}
+	satellites := make([]storj.NodeID, len(namespaces))
+	for i, namespace := range namespaces {
+		satellites[i], err = storj.NodeIDFromBytes(namespace)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return satellites, nil
+}
+
+// SpaceUsedBySatellite calculates disk space used for local piece storage in the given
+// satellite's namespace. Important note: this metric does not include space used by
+// piece headers, whereas storj/filestore/store.(*Store).SpaceUsedInNamespace() does
+// include all space used by the blobs.
+func (store *Store) SpaceUsedBySatellite(ctx context.Context, satelliteID storj.NodeID) (int64, error) {
+	var totalUsed int64
+	err := store.ForAllPieceIDsOwnedBySatellite(ctx, satelliteID, func(access StoredPieceAccess) error {
+		contentSize, statErr := access.ContentSize(ctx)
+		if statErr != nil {
+			store.log.Error("failed to stat", zap.Error(statErr), zap.String("pieceID", access.PieceID().String()), zap.String("satellite", satelliteID.String()))
```

Comment:
That's fair. As a counterpoint, though, this will always only be an approximation, because changes to the filesystem are probably ongoing while this runs. Maybe it would be better just to rename the whole function to "ApproximateSpaceUsedBySatellite" to avoid any confusion in the first place?

Reply:
That sounds good to me but I would also add a brief comment in the doc comment regarding why is approximated (errors and change ongoing).